osmorc/src/org/osmorc/maven/facet/OsmorcBndFacetImporter.java

```diff
@@ -0,0 +1,192 @@
+    addProperty(props, Constants.BUNDLE_NAME, mavenProject.getName());
+    props.computeIfAbsent(Constants.BUNDLE_VERSION, k -> new MavenVersion(mavenId.getVersion()).getOSGiVersion().toString());
+    addProperty(props, Constants.BUNDLE_DESCRIPTION, modelMap.get("description"));
+    addProperty(props, Constants.BUNDLE_VENDOR, modelMap.get("organization.name"));
+    addProperty(props, Constants.BUNDLE_SCM, modelMap.get("scm"));
+    addProperty(props, Constants.BUNDLE_DOCURL, modelMap.get("url"));
+
+    // add settings for reproducible build if Maven is configured for it
+    String outputTimestamp = mavenProject.getProperties().getProperty("project.build.outputTimestamp");
+    if (StringUtil.isNotEmpty(outputTimestamp)) {
+      props.putIfAbsent(Constants.NOEXTRAHEADERS, "true");
+      props.putIfAbsent(Constants.SNAPSHOT, "SNAPSHOT");
+    }
+
+    // Fix for IDEA-63242 - don't merge it with the existing settings, overwrite them
+    conf.importAdditionalProperties(props, true);
+
+    // Fix for IDEA-66235 - inherit jar filename from maven
+    String jarFileName = mavenProject.getFinalName() + ".jar";
+
+    // Fix for IDEA-67088, preserve existing output path settings on reimport.
+    switch (conf.getOutputPathType()) {
+      case OsgiOutputPath:
+        conf.setJarFileLocation(jarFileName, OutputPathType.OsgiOutputPath);
+        break;
+      case SpecificOutputPath:
+        String path = new File(conf.getJarFilePath(), jarFileName).getPath();
+        conf.setJarFileLocation(path, OutputPathType.SpecificOutputPath);
+        break;
+      default:
+        conf.setJarFileLocation(jarFileName, OutputPathType.CompilerOutputPath);
+    }
+  }
+
+  private static void addProperty(Map<String, String> headers, String headerName, String value) {
+    if (!StringUtil.isEmpty(value)) {
+      headers.putIfAbsent(headerName, value);
+    }
+  }
+
+  private String loadProjectProperties(Map<String, String> props, MavenProject mavenProject) throws IOException {
+    VirtualFile pomVirtualFile = mavenProject.getFile();
+
+    // get bnd-maven-plugin configuration from <plugin>/<execution> or fallback to <plugin>
+    Element configuration = getGoalConfig(mavenProject, BND_PLUGIN_GOAL);
+    if (configuration == null || configuration.getContent().isEmpty()) {
+      configuration = getConfig(mavenProject);
+    }
+
+    // configuration in <bnd> element
+    if (configuration != null) {
+      Element bndElement = configuration.getChild("bnd");
+      if (bndElement != null) {
+        LOG.debug("Using headers from bnd-element in " + mavenProject.getPath());
+        File pomFile = new File(pomVirtualFile.getPath());
+        UTF8Properties properties = new UTF8Properties();
+        properties.load(bndElement.getValue(), pomFile, null);
+        for (Map.Entry<?, ?> e : properties.replaceHere(pomFile.getParentFile()).entrySet()) {
+          props.put(e.getKey().toString(), e.getValue().toString());
+        }
```

Comment:
There's [one thing](https://github.com/bndtools/bnd/blob/master/maven/bnd-maven-plugin/README.md#configuration-parameters) with these properties:
> If the projects has a `bndfile` configuration property or a file in the default location `bnd.bnd`, then this configuration element is ignored.

The first case is easy to address, but that default location is tricky and probably doesn't worth bothering.

Reply:
I'm not too savvy with how IntelliJ mirrors the Maven model, but in the end I think it's impossible to exactly mirror the plugins' behavior. Mainly because the plugin traverses the pom hierarchy and parses each pom.xml individually, while IntelliJ presses all POMs into one flat model.

I can of course skip reading the properties here if a `<bndfile>` tag is found, but then it all instructions from `<bnd>` tag from a parent pom are also ignored because I can't distinguish where the properties came from.

I'd favor keeping it as is because if a bnd-file is present, regardless of via tag or default, the instructions in there take priority over the ones in the `<bnd>` tag. For example:

```
parent
|-- pom.xml:
|    <bnd>Bundle-DocUrl: https://example.org/parent</bnd>
|-- child
|    |--pom.xml:
|    |   <bnd>Bundle-DocUrl: https://example.org/child/from-tag</bnd>
|    |--bnd.bnd
|    |   Bundle-DocUrl: https://example.org/child/from-file
```

results in a manifest with `Bundle-DocUrl: https://example.org/child/from-file`. This is consistent with the Maven plugin.

However, in IntelliJ the `<bnd>` tag in the child overrides all instructions from the parent, while in Maven such a configuration is ignored and the parent is applied. As this is an invalid configuration per the plugin spec, I'm not sure if or how to deal with it given IntelliJ's simpliefied Maven model.

Reply:
I see, thanks. Okay, I believe it doesn't make sense to invest great efforts in exact replication of the plugin's behavior - standard cases seem to be covered, and for the rest there's a possibility to delegate build to Maven.